plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/findUsages/KotlinFindUsagesSupportFirImpl.kt

```diff
@@ -9,22 +9,26 @@ import com.intellij.psi.PsiElement
 import com.intellij.psi.PsiMethod
 import com.intellij.psi.PsiReference
 import com.intellij.psi.search.GlobalSearchScope
-import org.jetbrains.kotlin.idea.KotlinBundle
-import org.jetbrains.kotlin.idea.core.util.showYesNoCancelDialog
 import org.jetbrains.kotlin.analysis.api.analyseInModalWindow
 import org.jetbrains.kotlin.analysis.api.symbols.KtCallableSymbol
-import org.jetbrains.kotlin.lexer.KtTokens
-import org.jetbrains.kotlin.psi.*
 import org.jetbrains.kotlin.analysis.api.symbols.markers.KtNamedSymbol
 import org.jetbrains.kotlin.analysis.api.symbols.markers.KtSymbolWithKind
+import org.jetbrains.kotlin.idea.KotlinBundle
+import org.jetbrains.kotlin.idea.core.util.showYesNoCancelDialog
 import org.jetbrains.kotlin.idea.refactoring.CHECK_SUPER_METHODS_YES_NO_DIALOG
 import org.jetbrains.kotlin.idea.refactoring.formatPsiClass
+import org.jetbrains.kotlin.lexer.KtTokens
 import org.jetbrains.kotlin.name.SpecialNames
+import org.jetbrains.kotlin.psi.*
 import org.jetbrains.kotlin.psi.psiUtil.getElementTextWithContext
 
 class KotlinFindUsagesSupportFirImpl : KotlinFindUsagesSupport {
-    override fun isCallReceiverRefersToCompanionObject(element: KtElement, companionObject: KtObjectDeclaration): Boolean {
-        return false
+    override fun <T> analyseIfCallReceiverRefersToCompanionObject(
+        analyseScope: KtElement,
+        companionObject: KtObjectDeclaration,
+        block: (isCallReceiverRefersToCompanionObject: (KtElement) -> Boolean) -> T
+    ): T {
+        return block { false }
```

Comment:
Somehow we need to clarify/specify that some of `block` will be executed while some are not, depending on what `isCallReceiverRefersToCompanionObject` returns? For example, this `block { false }` looks like `block` will _not_ be executed no matter what. :\

Reply:
I am not sure I understand what you would like me to do here. This code is incomplete for now. Let me add some TODO to make this clear.

Reply:
Per Ilya's suggestion, I have removed this and make the entire `processCompanionObjectInternalReferences` abstract.